server/server.go

```diff
@@ -134,74 +150,148 @@ func (s *RPCServer) ListenAndServe() {
+		outputConn *amqp.Connection
+		inputCh    *amqp.Channel
+		outputCh   *amqp.Channel
+	)
+
+	inputConn, err := amqp.DialConfig(s.url, s.dialconfig)
 	if err != nil {
 		return err
 	}
+	defer inputConn.Close()
 
-	defer conn.Close()
-
-	inputCh, err := conn.Channel()
+	outputConn, err = amqp.DialConfig(s.url, s.dialconfig)
 	if err != nil {
 		return err
 	}
+	defer outputConn.Close()
 
+	inputCh, err = inputConn.Channel()
+	if err != nil {
+		return err
+	}
 	defer inputCh.Close()
 
+	outputCh, err = outputConn.Channel()
+	if err != nil {
+		return err
+	}
+	defer outputCh.Close()
+
+	// Setup a WaitGroup for use by consume(). This WaitGroup will be 0
+	// when all consumers are finished consuming messages.
+	consumersWg := sync.WaitGroup{}
+
+	// consumerTags is used when we later want to tell AMQP that we want to cancel our consumers.
+	consumerTags := []string{}
 	for _, binding := range s.bindings {
-		err := s.consume(binding, inputCh)
+		consumerTag, err := s.consume(binding, inputCh, consumersWg)
 		if err != nil {
 			return err
 		}
-	}
 
-	outputCh, err := conn.Channel()
-	if err != nil {
-		return err
+		consumerTags = append(consumerTags, consumerTag)
 	}
 
-	defer outputCh.Close()
+	// This WaitGroup will reach 0 when the responder() has finished sending all responses.
+	responderWg := sync.WaitGroup{}
+
+	go s.responder(outputCh, responderWg)
+
+	inputConnClosed := inputConn.NotifyClose(make(chan *amqp.Error))
+	outputConnClosed := outputConn.NotifyClose(make(chan *amqp.Error))
+
+	// Wait for graceful shutdown.
+	select {
+	case <-s.stopChan:
+		// 1. Tell amqp we want to shut down by cancelling all the consumers.
+		for _, consumerTag := range consumerTags {
+			err = inputCh.Cancel(consumerTag, false)
+			if err != nil {
+				return err
+			}
+		}
+
+		// 2. We've told amqp to stop delivering messages, now we wait for all
+		// the consumers to finish inflight messages.
+		consumersWg.Wait()
 
-	go s.responder(outputCh)
+		// 3. Close the responses chan and wait until the consumers are finished.
+		// We might still have responses we want to send.
+		close(s.responses)
+		responderWg.Wait()
 
-	err, ok := <-conn.NotifyClose(make(chan *amqp.Error))
+		// 4. We have no more messages incoming and we've sent all our responses.
+		inputConn.Close()
+		outputConn.Close()
+
+	// Check to see if the connections are closed without us calling .Close() first.
+	// This counts as an error even if AMQP thinks this was a graceful shutdown.
+	case err, ok := <-inputConnClosed:
+		if !ok {
+			return ErrUnexpectedConnClosed
+		}
+		return err
+	case err, ok := <-outputConnClosed:
+		if !ok {
+			// If we get here the connection was closed without calling Close().
+			// This counts as an error even if AMQP thinks this was a graceful shutdown.
+			return ErrUnexpectedConnClosed
+		}
+		return err
+	}
+
+	// Wait for the disconnection to complete after we have told the server to gracefully close.
+	err, ok := <-inputConnClosed
```

Comment:
You think it's too much to ensure an ok closure for output as well?

Reply:
Actually from the documentation of `Connection.Close`:

> After returning from this call, all resources associated with this connection,
including the underlying io, Channels, Notify listeners and Channel consumers
will also be closed.

This means that I won't have to do this waiting at all.

Reply:
And regarding the error that can be returned for `Close()`

> An error indicates that server may not have received this request to close but
the connection should be treated as closed regardless.

Although a user might want to log it in the application. We might need to expose the `Notify*` methods.